### PR TITLE
Implement Challenge Authorization

### PR DESCRIPTION
**Pull request by agunde406**
Also add the following to sawtooth_validator command line to change AuthType for network
sawtooth-validator -vv --network-auth challenge

Built on top of #822 Remove callback used for Gossip Peering

### CODE REVIEW

File: validator/packaging/validator.toml.example [hunk 1]
```diff
@@ -44,6 +44,12 @@ scheduler = 'serial'
 network_public_key = 'wFMwoOt>yFqI/ek.G[tfMMILHWw#vXB[Sv}>l>i)'
 network_private_key = 'r&oJ5aQDj4+V]p2:Lz70Eu0x#m%IwzBdP(}&hWM*'
 
+# The type of authorization that must be preformed for the different type of
```
Comment (jsmitchell) [inline validator/packaging/validator.toml.example]: s/preformed/performed/

File: validator/packaging/validator.toml.example [hunk 2]
```diff
@@ -44,6 +44,12 @@ scheduler = 'serial'
...
+# roles on the network. The different types of supported authorization types
```
Comment (jsmitchell) [inline validator/packaging/validator.toml.example]: s/types of//

File: validator/sawtooth_validator/networking/handlers.py [hunk 3]
```diff
@@ -224,16 +239,157 @@ def handle(self, connection_id, message_content):
...
 
 
+class AuthorizationChallengeRequestHandler(Handler):
+    def __init__(self, network):
+        self._network = network
+
+    def handle(self, connection_id, message_content):
+        """
+        If the connection wants to take on a role that requires a challenge to
+        be signed, it will request the challenge by sending the an
```
Comment (jsmitchell) [inline validator/sawtooth_validator/networking/handlers.py]: s/sending the an/sending an/

File: validator/sawtooth_validator/networking/handlers.py [hunk 4]
```diff
@@ -224,16 +239,157 @@ def handle(self, connection_id, message_content):
...
+        AuthorizationChallengeRequest to the validator it wishes to connect to.
+        The validator will send back a random payload that must be signed.
+        If the connection has not sent a ConnectionRequest or a the connection
```
Comment (jsmitchell) [inline validator/sawtooth_validator/networking/handlers.py]: s/or a the/or the/

File: validator/tests/test_authorization_handlers/test.py [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+                                        connect_message.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_CONNECTION_RESPONSE)
+
+    def test_connect_bad_role_type(self):
+        """
+        Test the ConnectHandler closes the connection if the role has an
+        unsupported roel type.
```
Comment (boydjohnson) [inline validator/tests/test_authorization_handlers/test.py]: roel -> role

File: validator/sawtooth_validator/networking/handlers.py [hunk 6]
```diff
@@ -224,16 +239,157 @@ def handle(self, connection_id, message_content):
...
+        has already recieved a AuthorizationChallengeResponse, an
```
Comment (jsmitchell) [inline validator/sawtooth_validator/networking/handlers.py]: s/recieved/received/ s/ a / an /

File: validator/sawtooth_validator/networking/handlers.py [hunk 7]
```diff
@@ -224,16 +239,157 @@ def handle(self, connection_id, message_content):
...
+        When the validator receives an AuthorizationChallengeSubmit message, it
+        will verify the public key against the signature. If the public key is
+        verified, the requested roles will be checked against the stored roles
+        to see if the public key is included in the policy. If the node’s
+        response is accepted, the node’s public key will be stored and the
+        requester may start sending messages for the approved roles.
+
+        If the requester wanted a role that is either not available on the
+        endpoint, the requester does not have access to one of the roles
+        requested, or the previous message was not a
```
Comment (jsmitchell) [inline validator/sawtooth_validator/networking/handlers.py]: s/not a/not an/

File: validator/tests/test_authorization_handlers/test.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+        handler_status = handler.handle("connection_id",
+                                        connect_message.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_CONNECTION_RESPONSE)
+
+    def test_connect_wrong_previous_message(self):
+        """
+        Test the ConnectHandler closes a connection has send a message before.
```
Comment (boydjohnson) [inline validator/tests/test_authorization_handlers/test.py]: Can this doc string be cleaned up? I can't tell what it means.

File: validator/packaging/validator.toml.example (hunk 1 above)
Comment (boydjohnson) [inline validator/packaging/validator.toml.example]: preformed -> performed

File: validator/tests/test_authorization_handlers/test.py [hunk 9]
```diff
@@ -0,0 +1,386 @@
...
+        handler_status = handler.handle("connection_id",
+                                        connect_message.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_CONNECTION_RESPONSE)
+
+    def test_authorization_trust_request(self):
+        """
+        Test the AuthorizationTrustRequestHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+            auth_trust_request.SerializeToString())
+
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_VIOLATION)
+
+    def test_authorization_challenge_request(self):
+        """
+        Test the AuthorizationChallengeRequestHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 11]
```diff
@@ -0,0 +1,386 @@
...
+            "connection_id",
+            auth_challenge_request.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_CHALLENGE_RESPONSE)
+
+    def test_authorization_challenge_request_bad_last_message(self):
+        """
+        Test the AuthorizationChallengeRequestHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 12]
```diff
@@ -0,0 +1,386 @@
...
+            "connection_id",
+            auth_challenge_request.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_VIOLATION)
+
+    def test_authorization_challenge_submit(self):
+        """
+        Test the AuthorizationChallengeSubmitHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+            "connection_id",
+            auth_challenge_submit.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_CHALLENGE_RESULT)
+
+    def test_authorization_challenge_submit_bad_last_message(self):
+        """
+        Test the AuthorizationChallengeSubmitHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 14]
```diff
@@ -0,0 +1,386 @@
...
+        AuthorizationViolation and closes the connection if the last message
+        was not AuthorizaitonChallengeRequest.
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/Authorizaiton/Authorization/

File: validator/tests/test_authorization_handlers/test.py [hunk 15]
```diff
@@ -0,0 +1,386 @@
...
+            "connection_id",
+            auth_challenge_submit.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_VIOLATION)
+
+    def test_authorization_challenge_submit_bad_signature(self):
+        """
+        Test the AuthorizationChallengeSubmitHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/tests/test_authorization_handlers/test.py [hunk 16]
```diff
@@ -0,0 +1,386 @@
...
+            "connection_id",
+            auth_challenge_submit.SerializeToString())
+        self.assertEqual(handler_status.status, HandlerStatus.RETURN_AND_CLOSE)
+        self.assertEqual(
+            handler_status.message_type,
+            validator_pb2.Message.AUTHORIZATION_VIOLATION)
+
+    def test_authorization_challenge_submit(self):
+        """
+        Test the AuthorizationChallengeSubmitHandler returns a
```
Comment (jsmitchell) [inline validator/tests/test_authorization_handlers/test.py]: s/returns a/returns an/

File: validator/sawtooth_validator/networking/handlers.py [hunk 17]
```diff
@@ -224,16 +239,157 @@ def handle(self, connection_id, message_content):
...
+
+    def handle(self, connection_id, message_content):
+        """
+        If the connection wants to take on a role that requires a challenge to
+        be signed, it will request the challenge by sending an
+        AuthorizationChallengeRequest to the validator it wishes to connect to.
+        The validator will send back a random payload that must be signed.
+        If the connection has not sent a ConnectionRequest or the connection
+        has already recieved an AuthorizationChallengeResponse, an
+        AuthorizatinViolation will be returned and the connection will be
```
Comment (jsmitchell) [inline validator/sawtooth_validator/networking/handlers.py]: Authorizatin

**Pull request merged**